File: apps/ios/Moodish/Moodish/Core/Auth/SessionStore.swift
import Foundation

/// Owns the personal-session bearer token and per-group-session access
/// tokens/participant ids. Backed by Keychain so tokens survive relaunch,
/// mirroring the web app's persisted cookie / localStorage identifiers.
@Observable
final class SessionStore {
    private static let sessionTokenKey = "moodish.session.token"
    private static let groupTokenPrefix = "moodish.group.token."
    private static let participantIdPrefix = "moodish.group.participant."

    private(set) var sessionToken: String?

    init() {
        sessionToken = KeychainStore.get(Self.sessionTokenKey)
    }

    func setSessionToken(_ token: String?) {
        sessionToken = token
        if let token {
            KeychainStore.set(token, forKey: Self.sessionTokenKey)
        } else {
            KeychainStore.remove(Self.sessionTokenKey)
        }
    }

    func groupAccessToken(sessionId: String) -> String? {
        KeychainStore.get(Self.groupTokenPrefix + sessionId)
    }

    func setGroupAccessToken(_ token: String, sessionId: String) {
        KeychainStore.set(token, forKey: Self.groupTokenPrefix + sessionId)
    }

    func participantId(sessionId: String) -> String? {
        KeychainStore.get(Self.participantIdPrefix + sessionId)
    }

    func setParticipantId(_ id: String, sessionId: String) {
        KeychainStore.set(id, forKey: Self.participantIdPrefix + sessionId)
    }

    func logout() {
        setSessionToken(nil)
    }
}

File: apps/moodish_android/lib/core/session_store.dart
import 'package:shared_preferences/shared_preferences.dart';
import 'package:flutter_secure_storage/flutter_secure_storage.dart';

/// Persists per-group-session access tokens and participant ids, keyed by
/// session id — mirrors the iOS app's Keychain-backed SessionStore.
class SessionStore {
  static const _secure = FlutterSecureStorage();
  static String _tokenKey(String sessionId) => 'moodish.group.token.$sessionId';
  static String _participantKey(String sessionId) => 'moodish.group.participant.$sessionId';

  Future<String?> groupAccessToken(String sessionId) async {
    final prefs = await SharedPreferences.getInstance();
    final token = await _secure.read(key: _tokenKey(sessionId)) ?? prefs.getString(_tokenKey(sessionId));
    if (token != null) await _secure.write(key: _tokenKey(sessionId), value: token);
    await prefs.remove(_tokenKey(sessionId));
    return token;
  }

  Future<void> setGroupAccessToken(String sessionId, String token) async {
    await _secure.write(key: _tokenKey(sessionId), value: token);
  }

  Future<String?> participantId(String sessionId) async {
    final prefs = await SharedPreferences.getInstance();
    return prefs.getString(_participantKey(sessionId));
  }

  Future<void> setParticipantId(String sessionId, String id) async {
    final prefs = await SharedPreferences.getInstance();
    await prefs.setString(_participantKey(sessionId), id);
  }
}
